### ISSUE TITLE
Support for Multi-Factor Auth?

### ISSUE THREAD

**Issue by rscotten** [firebase/firebaseui-web#782]
Hello,

I just wired up Multi-Factor authentication for my password auth, and it works great.  But when I try to log in via Google Auth through FirebaseUI, I get the following popup message:

<img width="268" alt="Screen Shot 2020-11-15 at 3 23 47 PM" src="https://user-images.githubusercontent.com/8073294/99199813-9bd48900-2756-11eb-8c97-c043337c0f56.png">

says "Proof of ownership of a second factor is required to complete sign-in."

The callbacks `signInFailure` and `signInSuccessWithAuthResult` are not called, so I can't handle this error and redirect to my confirmation code form.

Also, I can't find any examples online on how to use MFA with FirebaseUI.

Question 1: Does Firebase UI support MFA?
Question 2: If so, is there a canonical example on how to make it work?
Question 3: If Question #1 is a no, will it be or can it please be supported?

**Comment by bojeil-google** [comment]
Hey @rscotten, we don't yet support multi-factor auth in FirebaseUI.
We plan a phased approach for support this capability to unblock developers like yourself:

1. Provide a callback to plug in your own 2nd factor handler.
2. Provide our FirebaseUI built-in implementation for the 2nd factor handler.

**Comment by rscotten** [comment]
@bojeil-google Thanks for the quick response. Your phase 1 would be greatly appreciated!

**Comment by TaLoN1x** [comment]
looking forward for this feature. It's a blocker for my project

**Comment by rscotten** [comment]
@TaLoN1x MFA exists now.

https://firebase.google.com/docs/reference/android/com/google/firebase/auth/MultiFactor

**Comment by ferrywlto** [comment]
Same here. Is there a way for us to catch the `auth/multi-factor-auth-required` error somewhere at least? Thank you.

**Comment by bojeil-google** [comment]
We will expose a callback to plug in your own 2nd factor handler. This will be triggered on `auth/multi-factor-auth-required`. We are busy working on another feature at the moment. This should be next on our list.

**Comment by munderwoods** [comment]
I am also blocked by this. Any ETA for this? Thanks.

**Comment by SUVOLewis** [comment]
Any update on this? I'm currently still unable to handle the `verifyPassword: Proof of ownership of a second factor is required to complete sign-in.` error, so I can't divert the flow to any of my custom elements.

**Comment by munderwoods** [comment]
> Any update on this? I'm currently still unable to handle the `verifyPassword: Proof of ownership of a second factor is required to complete sign-in.` error, so I can't divert the auth flow to any of my custom elements.

I've ended up implementing authy then using claims on the user's google object to identify if they have 2fa. I am hoping that this issue is resolved before this app launches. @bojeil-google did say that it "should be next on (their) list" on Dec 23 so I have to imagine some progress has been made. 3 months for google? Could go either way I suppose.

**Comment by rscotten** [comment]
@bojeil-google It's been nine months since I originally posted this.  The industry that we serve just had a major hacking (allegedly by Russians and the FBI is involved) and our clients are now motivated to enable MFA.   We'd really appreciate if this can be given priority attention (what's more important than the security of our data?).  We look silly when we have to explain to our customers that MFA works for email and not OAuth.

**Comment by akauppi** [comment]
@rscotten Would one way be to abandon `firebaseui-web` completely and make a true community based web auth UI? I’m in.

**Comment by rowbot-weisguy** [comment]
I ended up having to re-implement a basic UI, which I'm sure is less battle-tested and more bug-prone than this project, in order to enable MFA for our customers. 

It appears the Firebase team is struggling to get / manage resources and treats their UI as second-rate to the API. Perhaps because they don't charge for use of the UI itself. That said, I don't know if it's wise to start a community project tightly coupled to Firebase auth. If their team can't effectively support their own product, should the community start doing free work for them? I appreciate the spirit of open source, but this smells different. I wish Firebase would treat their UI as part of the product that people pay for.

**Comment by nguyenduyhust** [comment]
Any new information about this feature? Or is there any workaround for it?

**Comment by rscotten** [comment]
@bojeil-google is this still on the roadmap?

**Comment by gnahzknarf** [comment]
this is hopeless. is there any workaround for it?

**Comment by iamjuanguillermo** [comment]
Ago/24 -  Hi all! Any update on this?

**Comment by kartikwatwani** [comment]
Need update for MFA.

**Comment by HSCOGT** [comment]
Is there any workaround, given that it doesn’t look like it will ever be integrated?

**Comment by dackers86** [comment]
Thank you for taking the time to submit this feature request.

We are happy to let you know that a resolution is already planned and on the roadmap for our upcoming new version. 

We are actively working on the new SDK and will be sure to update this issue once we have a public beta or a release candidate that includes this functionality. Please watch the main repository for the latest release.

In the meantime, we will keep this feature request open for tracking.

**Comment by russellwheatley** [comment]
v7 beta has been released with MFA support. We no longer support v6. if you find a bug on v7, feel free to file a new issue.
